atlp-rwanda/dynamites-ecomm-be
# Manage product

## What does this PR do?
This pull request adds the product management feature to our app. It defines the products and the category models, categories are used to group products which means there is a relationship between product and category model. 
The PR also allows vendors to perform all CRUD Operations on Products and Categories such means a vendor can add, update, read, and delete products and categories

## Description of Task to be completed?
### Create
- [x]  An Endpoint for sellers to create categories and products
- [x] Validate the requesting user's role as a seller
- [x]  Validate the request body (price, name, category, expiry date, bonus, etc.)
- [x] Check if the category exists in the database, if not add it to the database to be used by all sellers
- [x]  Check if the target a product exists in the database
- [x]  If the item does not exist, add it to the database
- [x]  Error handling and appropriate response messages for validation failures
- [x]  Input sanitization
- [x]  Ensuring the product has a reference to the associated seller

### Read
- [x]  An API endpoint for retrieving all products and one product based on Id
- [x]  An API endpoint for retrieving all categories and one category based on Id

### Update
- [x]  Implement an API endpoint for updating a seller's product details
- [x]  Implement an API endpoint for updating category details
- [x]  Validate the requesting user's role as a seller
- [x]  Retrieve the product from the database based on the provided product ID
- [x]  Update the product  and category details in the database
- [x]  Implement error handling and appropriate response messages for validation failures
- [x]  Implement input sanitization

### Delete
- [x]  Implement an API endpoint for deleting a product or category based on ID
- [x]  Validate the requesting user's role as a seller
- [x]  Retrieve the product or category from the database based on the provided product or category ID
- [x]  Delete the product or category from the database
- [x]  Implement error handling and appropriate response messages for validation failures
- [x]  Implement input sanitization

## How to Test
Clone the Branch, install all dependencies and run this command: npm run dev

First Register as a vendor if you do not have an account(if you have account skip to number 3)
POST: http://localhost:3000/api/v1/register

Verify your account by providing your code from your email
POST: http://localhost:3000/api/v1/confirm

Login and get Token from email
POST: http://localhost:3000/api/v1/login

Verify 2FA
POST: http://localhost:3000/api/v1/verify2FA/:userId

Add a category
POST: http://localhost:3000/api/v1/category

Add a product
POST: http://localhost:3000/api/v1/product

Update a category
PUT: http://localhost:3000/api/v1/category/{categoryId}

Update a product
PUT: http://localhost:3000/api/v1/product/:productId

Delete a product
DELETE: http://localhost:3000/api/v1/product/{productId}

Delete a Category
DELETE: http://localhost:3000/api/v1/category/{categoryId}

Get all products and one product by Id
GET: http://localhost:3000/api/v1/product/
GET: http://localhost:3000/api/v1/product/{productId}

Get all categories and one category by Id
GET: http://localhost:3000/api/v1/category
GET: http://localhost:3000/api/v1/category/{categoryId}

##  Test Screenshots
![56c0a0c7-4096-427c-8ef0-4f42e4822c02](https://github.com/atlp-rwanda/dynamites-ecomm-be/assets/51314498/fd989502-5343-42a0-bd7f-e83553258b8a)

## Swagger Screenshot
![WhatsApp Image 2024-05-16 at 18 10 09](https://github.com/atlp-rwanda/dynamites-ecomm-be/assets/51314498/ebdd16e9-1791-425f-b3cf-3d8b2fd45589)

## Any background context you want to provide?
Right know there are some features that have been not added yet such as Currency and Seller Collection features, we worked on whats important for the team member who are depending on our PR to start working, but we are looking for a way to implement these feature after merging.
